pkg/actor/decommission.go

```diff
@@ -0,0 +1,166 @@
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actor
+
+import (
+	"context"
+	"fmt"
+
+	api "github.com/cockroachdb/cockroach-operator/api/v1alpha1"
+	"github.com/cockroachdb/cockroach-operator/pkg/clustersql"
+	"github.com/cockroachdb/cockroach-operator/pkg/condition"
+	"github.com/cockroachdb/cockroach-operator/pkg/database"
+	"github.com/cockroachdb/cockroach-operator/pkg/features"
+	"github.com/cockroachdb/cockroach-operator/pkg/kube"
+	"github.com/cockroachdb/cockroach-operator/pkg/resource"
+	"github.com/cockroachdb/cockroach-operator/pkg/scale"
+	"github.com/cockroachdb/cockroach-operator/pkg/utilfeature"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func newDecommission(scheme *runtime.Scheme, cl client.Client, config *rest.Config) Actor {
+	return &decommission{
+		action: newAction("decommission", scheme, cl),
+		config: config,
+	}
+}
+
+// decommission performs the initialization of the new cluster
+type decommission struct {
+	action
+
+	config *rest.Config
+}
+
+func (d decommission) Handles(conds []api.ClusterCondition) bool {
+	return condition.False(api.NotInitializedCondition, conds) && utilfeature.DefaultMutableFeatureGate.Enabled(features.Decommission)
+}
+
+func (d decommission) Act(ctx context.Context, cluster *resource.Cluster) error {
+	log := d.log.WithValues("CrdbCluster", cluster.ObjectKey())
+	log.Info("check decommission oportunities")
+
+	stsName := cluster.StatefulSetName()
+
+	key := kubetypes.NamespacedName{
+		Namespace: cluster.Namespace(),
+		Name:      stsName,
+	}
+	ss := &appsv1.StatefulSet{}
+	if err := d.client.Get(ctx, key, ss); err != nil {
+		log.Info("decommission failed to fetch statefulset")
+		return kube.IgnoreNotFound(err)
+	}
+	status := &ss.Status
+
+	if status.CurrentReplicas == 0 || status.CurrentReplicas < status.Replicas {
+		log.Info(" decommission statefulset does not have all replicas up")
+		return NotReadyErr{Err: errors.New("decommission statefulset does not have all replicas up")}
+	}
+
+	replicas := uint(cluster.Spec().Nodes)
+	if replicas < 3 {
+		log.Info("We cannot decommission if there are less than 3 nodes", "nodes", replicas)
+		return errors.New("decommission with less than 3 nodes is not supported")
+	}
+	log.Info("replicas decommisioning", "status.CurrentReplicas", status.CurrentReplicas, "expected", cluster.Spec().Nodes)
+	if status.CurrentReplicas > cluster.Spec().Nodes {
```

Comment:
nit: if there's a short circuit case prefer to handle that with an early return for readability.

```go
if status.CurrentReplicas < cluster.Spec().Node {
    // nothing to do
    return nil
}
```